src/lib/db.ts

```diff
@@ -383,11 +404,10 @@ export class DB {
             });
         })
         .then((newDocument) => {
-          return this.openProblemOrPlanningDocument(type, newDocument.documentKey);
-        })
-        .then((newDocument) => {
-          documents.add(newDocument);
-          return newDocument;
+          // reset the (presumably null) promise for this document type
+          documents.addRequiredDocumentPromises([ProblemDocument]);
+          // return the promise, which will be resolved by the DB listener
+          return documents.requiredDocuments[ProblemDocument].promise;
```

Comment:
The prior code called `openProblemOrPlanningDocument` which creates a new document instance. Instead, we return the shared promise from the `documents` store.

Reply:
>It is a little confusing that a document was created above which I guess is a MST document

One of the things that would help is developing a naming convention that distinguished MST documents from firebase documents, but at this point most of these methods are creating firebase documents and then the MST documents are created by listeners.

Reply:
That argument wasn't very convincing -- even to me 😉 -- so let's try this one. There are two players involved here: the DB module, which is responsible for creating/updating firebase documents, and the DB listeners which are responsible for reading firebase documents and creating MST documents for them. The DB module always awaits the promise before it does anything, so the DB listeners are guaranteed to make the first move, which is to either resolve it with an appropriate document or with null if there are no appropriate documents. If the promise is resolved with null then the DB module will create the required firebase document, which the DB listeners will then respond to by creating the corresponding MST document. If we didn't reset the promise when we create the firebase document, then it would stay resolved to null indefinitely. Having the DB module reset the promise is what allows the DB listeners to resolve it with the newly-created MST document. If there were multiple clients with the ability to create the firebase document then there would be a race condition, but since only the DB module ever creates the firebase documents and only the DB listeners ever create the MST models at this point, the promise serves as an appropriate semaphore between them.

Reply:
That is helpful and makes sense.

The case that I was thinking about was this:
Some component that is intended to render info about a document on initialization waits on the document promise.  In this case it would only wait on the first promise and when it resolves to null this component would just render null.  Unless it knew to re-wait after receiving a null result, it would never get the actual document.

But I think this case doesn't exist. These promises are just used internally.  So I'd guess any component like that should be rendering a document from the MST document store. So it will initially not have a document, but then when the document finally gets created or loaded now this MST document will be there and the component will render it.